tests/MongoDB.Driver.Core.Tests/Core/Authentication/MongoAWSAuthenticatorTests.cs

```diff
@@ -0,0 +1,188 @@
+            
+            AwsSignatureVersion4.SignRequest(
+                dateTime,
+                credential.Username,
+                credential.GetInsecurePassword(),
+                null,
+                serverNonce,
+                host,
+                out var authHeader,
+                out var timestamp);
+
+            var dateTimeProviderMock = new Mock<IClock>();
+            dateTimeProviderMock.Setup(x => x.UtcNow).Returns(dateTime);
+
+            var randomByteGeneratorMock = new Mock<IRandomByteGenerator>();
+            randomByteGeneratorMock.Setup(x => x.Generate(It.IsAny<int>())).Returns(clientNonce);
+
+            var expectedClientFirstMessage = new BsonDocument()
+                .Add("r", new BsonBinaryData(clientNonce))
+                .Add("p", new BsonInt32('n'));
+            var expectedClientSecondMessage = new BsonDocument()
+                .Add("a", authHeader)
+                .Add("d", timestamp);
+
+            var serverFirstMessage = new BsonDocument()
+                .Add("s", new BsonBinaryData(serverNonce))
+                .Add("h", new BsonString(host));
+            var saslStartReply = MessageHelper.BuildReply<RawBsonDocument>(RawBsonDocumentHelper.FromJson(
+                $"{{conversationId: 1, done: false, payload: BinData(0,\"{ToBase64(ToBytes(serverFirstMessage))}\"), ok: 1}}"));
+            var saslContinueReply = MessageHelper.BuildReply<RawBsonDocument>(RawBsonDocumentHelper.FromJson(
+                "{conversationId: 1, done: true, payload: BinData(0,\"\"), ok: 1}"));
+
+            var subject = new MongoAWSAuthenticator(credential, null, randomByteGeneratorMock.Object, dateTimeProviderMock.Object);
+
+            var connection = new MockConnection(serverId);
+            connection.EnqueueReplyMessage(saslStartReply);
+            connection.EnqueueReplyMessage(saslContinueReply);
+
+            var expectedRequestId = RequestMessage.CurrentGlobalRequestId + 1;
+
+            Exception exception;
+            if (async)
+            {
+                exception = Record.Exception(() => subject.AuthenticateAsync(connection, connectionDescription, CancellationToken.None).GetAwaiter().GetResult());
+            }
+            else
+            {
+                exception = Record.Exception(() => subject.Authenticate(connection, connectionDescription, CancellationToken.None));
+            }
+            exception.Should().BeNull();
+
+            SpinWait.SpinUntil(() => connection.GetSentMessages().Count >= 2, TimeSpan.FromSeconds(5)).Should().BeTrue();
+
+            var sentMessages = MessageHelper.TranslateMessagesToBsonDocuments(connection.GetSentMessages());
+            sentMessages.Count.Should().Be(2);
+
+            var actualRequestId0 = sentMessages[0]["requestId"].AsInt32;
+            var actualRequestId1 = sentMessages[1]["requestId"].AsInt32;
+            actualRequestId0.Should().BeInRange(expectedRequestId, expectedRequestId + 10);
```

Comment:
Hm I wonder if this pattern was meant for "live" connections instead of "mocked" connections.  We can keep it, or we can change it to use +1: I defer to Mikalai.